BlobVM (#871)
ava-labs/avalanche-docs: docs/subnets/create-a-vm-blobvm.md

BlobVM has the following components to handle the tasks from transaction to block acceptance:

- **Transaction** - Transaction structure, initialization, execution etc.
- **Mempool** - Heap (min and max) for keeping pending transactions locally on the node.
- **Network** - Implements gossip of transactions from mempool to the network.
- **Block** - Handles block creation, initialization, verification, parsing, etc.
- **Builder** - Returns the built preferred block by including transactions from the mempool.
- **Block Builder** - Initiates the gossip of new transactions, notifies the engine that new blocks are ready to be built, and also handles next block generation notification.
- **Storage** - Stores and retrieves data of the chain's state.
- **Virtual Machine** - Entry point for all the components to orchestrate them according to the consensus engine and API requests.
- **Service** - API handlers for interacting with VM and initialized chain.
- **Factory** - For creating new instances of the Virtual Machine.

## Transaction Lifecycle in BlobVM

> A Virtual Machine exposes APIs or handlers for users to make direct RPC to service or use the client to interact with the service. Every change on a chain happens through blocks and more specifically, transactions. A VM handles transactions internally since the consensus engine only cares about the block. Let's see, how a transaction goes through the network to update the chain's state:

Reviewer: what does the client in `use the client` refer to?

Author: I should remove the `client` reference, as this is not a requirement for BlobVM.

- User calls `client.IssueRawTx` or directly make RPC to `service`
- `service.IssueRawTx()` is called using handlers
  - Receive transaction bytes as arguments
  - Unmarshal bytes into a transaction object
  - Initialize transaction object with message digest, txID, etc.
  - Submit transaction to VM
- Transaction is submitted to virtual machine - `vm.Submit()`
  - Get the preferred (last accepted) block
  - Get the execution context of the preferred block that includes
    - Recent TxIDs (Txns in the lookback window e.g last 10s as defined in Genesis)
    - Recent BlockIDs
    - Recent Price and Cost
    - Next Price, Next Cost, etc.
  - Execute transaction locally with execution context, dummy database, and block
  - Add valid transaction to mempool - `mempool.newTxs`
    - Call `mempool.addPending()` to signal VM to build block for the newly added tx
- Gossip new transactions from `mempool.newTxs` at regular intervals

> - Signals consensus engine to build blocks out of pending transactions in the mempool

Reviewer: Please clarify who `Gossips` and `Signals`.

- ProposerVM delays the request until it is the node's turn to propose a block
- The consensus engine calls `vm.BuildBlock()` to get the block from VM
- Engine calls `block.Verify()` method

> - Successfully verified blocks have gossiped within the network for consensus

Reviewer: ```suggestion
- Successfully verified blocks have been gossiped within the network for consensus
```

- Blocks containing transactions are accepted or rejected according to the consensus results
- Accepted blocks and related data are committed to the node's database

## Coding the Virtual Machine

We have divided the components into 3 packages. We will be looking at each of these files and learning about their functions.

> - **vm**

Reviewer: Link to the directory on github project will be helpful.

- [`FeeUnits()`](https://github.com/ava-labs/blobvm/blob/master/chain/base_tx.go#L59) returns the fee units this transaction will consume.
- [`LoadUnits()`](https://github.com/ava-labs/blobvm/blob/master/chain/base_tx.go#L63) same as fee units that return the units this transaction will consume.
- [`ExecuteBase()`](https://github.com/ava-labs/blobvm/blob/master/chain/base_tx.go#L46) executes the basic checks for a transaction like magic number validation, minimum transaction price, etc.

```go
func (b *BaseTx) ExecuteBase(g *Genesis) error {
	if b.BlockID == ids.Empty {
		return ErrInvalidBlockID
	}
	if b.Magic != g.Magic {
		return ErrInvalidMagic
	}
	if b.Price < g.MinPrice {
		return ErrInvalidPrice
	}
	return nil
}
```

- [`Execute()`](https://github.com/ava-labs/blobvm/blob/master/chain/unsigned_tx.go#L34) executes the specific check for a transaction and may perform state change on the database instance provided as an argument. Each type of transaction should implement its own execute method. For eg. `TransferTx` execute balance modification i.e. add transfer amount to the receiver and deduct the same amount from the sender.

> A transaction is executed 2 times. Before [including](https://github.com/ava-labs/blobvm/blob/master/vm/vm.go#L428) it in mempool and during [verification](https://github.com/ava-labs/blobvm/blob/master/chain/block.go#L213) of the block containing this transaction. The database for the former is aborted as this is just a local execution for validating transactions before gossiping. Whereas the database for the latter is committed after the block is accepted by the network.

Reviewer: This paragraph is not easy to understand. I would suggest to rewrite it.

Author: Added a few clarifications.

# How to Build a Complex VM

This is part of a series of tutorials for building a Virtual Machine (VM):

- [Introduction to Virtual Machines](./introduction-to-vm.md)
- [How to Build a Simple VM](./create-a-vm-timestampvm.md)
- How to Build a Complex VM (this article)

## Introduction

> In this tutorial, we will learn all about how we can build a virtual machine by referencing [BlobVM](https://github.com/ava-labs/blobvm). It's a Virtual Machine that can be used to instantiate key-value blockchains for storing files like images, videos, etc. in an efficient way.

Reviewer 2: I don't think it's true that BlobVM will scale well for images/videos, so we should probably remove those references.

Reviewer 3: I removed this reference and just said generic k/v storage.

Totally an aside but I think it kind of does scale fine but it would depend on the use-case/if you consider gas prices an issue in scaling :p

- [`FeeUnits()`](https://github.com/ava-labs/blobvm/blob/master/chain/base_tx.go#L59) returns the fee units this transaction will consume.
- [`LoadUnits()`](https://github.com/ava-labs/blobvm/blob/master/chain/base_tx.go#L63) same as fee units that return the units this transaction will consume.
- [`ExecuteBase()`](https://github.com/ava-labs/blobvm/blob/master/chain/base_tx.go#L46) executes the basic checks for a transaction like magic number validation, minimum transaction price, etc.

```go
func (b *BaseTx) ExecuteBase(g *Genesis) error {
	if b.BlockID == ids.Empty {
		return ErrInvalidBlockID
	}
	if b.Magic != g.Magic {
		return ErrInvalidMagic
	}
	if b.Price < g.MinPrice {
		return ErrInvalidPrice
	}
	return nil
}
```

- [`Execute()`](https://github.com/ava-labs/blobvm/blob/master/chain/unsigned_tx.go#L34) executes the specific check for a transaction and may perform state change on the database instance provided as an argument. Each type of transaction should implement its own execute method. For example, `TransferTx` execute balance modification, i.e. add transfer amount to the receiver and deduct the same amount from the sender.

> A transaction is executed 2 times. Before [including](https://github.com/ava-labs/blobvm/blob/master/vm/vm.go#L428) it in mempool and during [verification](https://github.com/ava-labs/blobvm/blob/master/chain/block.go#L213) of the block containing this transaction. As mentioned earlier, verification of a block happens before gossiping it within the network for consensus.

Reviewer: ```suggestion
A transaction is executed 2 times: before [including](https://github.com/ava-labs/blobvm/blob/master/vm/vm.go#L428) it in mempool and during [verification](https://github.com/ava-labs/blobvm/blob/master/chain/block.go#L213) of the block containing this transaction. As mentioned earlier, verification of a block happens before gossiping it within the network for consensus.
```

> # Create a Blob Virtual Machine (VM)

Reviewer 3: High-level: 

1. I think we're going into too much detail in some parts of this doc. We should make a clear distinction between what must be implemented as part of a VM (e.g the VM's public interface), vs what was optionally added into the BlobVM and why.
2. We should run this through a grammar checker to get rid of awkward phrasing/run-on sentences/etc. I've used Grammarly and it's pretty nice. There might be some VS-Code plugin for it (I don't use VS code but I think the majority of the team does. Saw [znck](https://marketplace.visualstudio.com/items?itemName=znck.grammarly)  and [LTeX](https://marketplace.visualstudio.com/items?itemName=valentjn.vscode-ltex) looked like good candidates on google). 
3. Should we avoid duplicating common stuff in our [vm doc](https://docs.avax.network/subnets/create-a-virtual-machine-vm#introduction) and only put blob-vm specific stuff here? We can probably safely assume the reader has read that doc as a prereq, or post a link at the top of this doc to let the reader know they should be familiar with that doc before reading this.

Author: Initially, we were planning to replace the timestamp tutorial, that's why you were seeing duplications.

## Introduction

Virtual Machine is a blueprint for a blockchain and a blockchain instantiates from VM, very much similar to a class-object relationship. Issuance of transactions, transaction types, block structure, block building algorithm, keeping pending transactions in the mempool, gossiping transactions to the connected nodes, etc. are defined in a virtual machine.

> Virtual Machines can be broken down into components like **Transaction**, **Mempool**, **Network**, **Block**, etc. Each component works together to:

Reviewer 3: Avoid using "etc", since this is a tutorial we shouldn't presume knowledge from the reader. Also I don't think a VM's interface requires us to implement a mempool (even though it is a common abstraction used).

Author: "etc" will be explained later in the tutorial

- Define the representation of the blockchain's state
- Represent the operations in that state
- Apply the operations in that state

## BlobVM

In this tutorial, we will learn all about virtual machines and how we can build them by taking references from [BlobVM](https://github.com/ava-labs/blobvm). It's a virtual machine that can be used to instantiate key-value blockchains for storing files like images, videos, etc. in an efficient way.

> Blobs are small chunks of data. In BlobVM, we divide a file into small blobs and store them as a key-value pair. The key of these small chunks of a file is linked together as the children of the root. This tree is also stored as JSON data against the root's key.

Reviewer 3: Diagram might be helpful here imo

## How to Use a VM

VM modules are served as a binary plugin to the `avalanchego` software. The nodes that want to power their chains using a particular VM, must have the built binary in their `$GOPATH/src/github.com/ava-labs/avalanchego/build/plugins/` (if they haven't built from source), and `$SOURCE_DIR/build/plugins/` if they are building from source. This can also be customized (see [here](../nodes/maintain/avalanchego-config-flags.md#--build-dir-string) for more details). There could be multiple VM plugins in this directory.

## High-Level Overview

Virtual Machines are created as a module, whose binary is registered by a node running `avalanchego`, against the **vmID** (binary file name must be vmID). VMID is a user-defined string that is zero-extended to a 32-byte array and encoded in CB58.

We can build multiple chains with the registered VM. A VM is registered, only when its built binary is present at the required location, as explained [here](#how-to-use-a-vm).

### APIs for a VM

> The VM exposes 3 things for interaction:

Reviewer 3: Wording here kind of suggests to me that clients are required as part of the VM interface, which isn't the case

- **Static and Non-Static Handlers (API)** for making requests to VM methods
- **Service** for implementing handlers or APIs
- **Client** for making calls to the service's APIs

### What are Handlers

Handlers serve the response for the incoming HTTP requests. Handlers can also be wrapped with **gRPC** for efficiently making calls from other services such as `avalanchego`. Handler helps in creating APIs. VM implements 2 kinds of handlers:

- **Non-Static Handlers** - They help in interacting with blockchains instantiated by the VM. The API's endpoint will be different for different chains. `/ext/bc/[chainID]`

> - **Static Handlers** - They help in directly accessing VM. These are optional for reasons such as parsing genesis bytes required to instantiate new blockchains. `/ext/vm/[vmID]`

Reviewer 3: its a bit unclear what the difference between these two are from reading this

### Registering a VM

While [registering a VM](https://github.com/ava-labs/avalanchego/blob/master/vms/manager.go#L93), we store its **factory** against the **vmID**. As the name suggests, a VM factory can create new VM instances. Using the VM's instance, we can initialize it and can create a new chain. Note that, instantiating a VM is different than initializing it. We initialize the instance of a VM with parameters like genesis, database manager, etc. to create a functional chain. A functional chain doesn't mean the node will start issuing blocks and processing transactions. This will only happen once the node is bootstrapped.

```go
// Registering a factory inside avalanchego
func (m *manager) RegisterFactory(vmID ids.ID, factory Factory) error {
    ...
    m.factories[vmID] = factory
    ...
}
```

> ### Instantiating a VM

Reviewer 3: Feels like we're duplicating with the `timestampvm` tutorial in a lot of places in the doc [here](https://docs.avax.network/subnets/create-a-virtual-machine-vm#factory). Maybe we should break this + the timestampvm apart to have three docs? 
1. What needs to be implemented for a VM, explains each part of the vm interface and what it does
2. Timestamp tutorial, just references (1) and explains additional stuff we built on top to support timestampvm-specific features
3. Blob tutorial, similarly references (1) and explains additional stuff we built on top to support blobvm-specific features.

cc: @cgcardona @aaronbuchwald

Author: Hi @joshua-kim. Thanks for this suggestion. We are moving ahead with this 3-tutorial VM series. I have made a recent commit to support this. Please have a look. Thank you.

Each VM has a **factory** that is capable of creating new VM instances. A VM can be initialized as a chain along with handlers for accessing it, only when we have the VM's instance. The factory's `New` method shown below, returns the VM's instance to its caller in `avalanchego`. It's generally in the [`factory.go`](https://github.com/ava-labs/blobvm/blob/master/factory.go) file of the VM.

```go
// Returning a new VM instance from VM's factory
func (f *Factory) New(*snow.Context) (interface{}, error) { return &vm.VM{}, nil }
```

> A VM's functionality is exposed to `avalanchego` by registering its factory and [instantiating](https://github.com/ava-labs/avalanchego/blob/master/chains/manager.go#L399) a new VM using this factory whenever a chain is required to be built.

Reviewer 3: IMO we can also remove this

```go
if err := vm.Initialize(
    ctx.Context,
    vmDBManager,
    genesisData,
    chainConfig.Upgrade,
    chainConfig.Config,
    msgChan,
    fxs,
    sender,
);
```

You can refer to the [implementation](https://github.com/ava-labs/blobvm/blob/master/vm/vm.go#L92) of `vm.initialize` in the BlobVM repository.

### Interaction

We can use the API handlers to issue transactions, query chain state, and whatever functionalities are being provided by the VM.

## Interfaces

The main task of a VM is to issue blocks to the consensus engine whenever requested. It is the VM's responsibility to handle user transactions, and include them into new blocks, mempool handling, and database handling.

The consensus engine will just request the VM for new blocks whenever there is a signal (from VM) and according to other [congestion control mechanisms](https://github.com/ava-labs/avalanchego/blob/master/vms/proposervm/README.md), verify the block, gossip the block within the network for consensus, and request the VM to accept or reject the block.

Every VM should implement the following interfaces:

### `block.ChainVM`

> To reach a consensus on linear blockchains (as opposed to DAG blockchains), Avalanche uses the Snowman consensus engine. To be compatible with Snowman, a VM must implement the `block.ChainVM` interface, which can be accessed from [AvalancheGo repository](https://github.com/ava-labs/avalanchego/blob/v1.7.4/snow/engine/snowman/block/vm.go).

Reviewer 3: In the vein of the previous comment, I feel like we need to do some sort of doc refactor to eliminate duplicate info with https://docs.avax.network/subnets/create-a-virtual-machine-vm#blockchainvm.

# Create a Blob Virtual Machine (VM)

> [toc]

Reviewer: This is not needed.

Virtual Machines can be broken down into components like **Transaction**, **Mempool**, **Network**, **Block**, etc. Each component works together to -
* Define the representation of the blockchain's state
* Represent the operations in that state
* Apply the operations in that state

## BlobVM

In this tutorial, we will learn all about virtual machines and how we can build them by taking references from **[BlobVM](https://github.com/ava-labs/blobvm)**. It's a virtual machine that can be used to instantiate key-value blockchains for storing files like images, videos, etc. in an efficient way.

Blobs are small chunks of data. In BlobVM, we divide a file into small blobs and store them as a key-value pair. The key of these small chunks of a file is linked together as the children of the root. This tree is also stored as JSON data against the root's key.

> ## How to use a VM

Reviewer: Title case for all titles, please

> VM modules are served as a binary plugin to the `avalanchego` software. The nodes that want to power their chains using a particular VM, must have the built binary in their avalanchego's `build/plugins` directory. There could be multiple VM plugins in this directory.

Reviewer 3: This is dependent on the user's configuration. We can point them to the docs ([ref])(https://docs.avax.network/nodes/maintain/avalanchego-config-flags#build-directory), but this will default to `$GOPATH/src/github.com/ava-labs/avalanchego/build/plugins/` (if they haven't built from source), and `$SOURCE_DIR/build/plugins/` if they are building from source.

## High-Level Overview

> Virtual Machines are created as a module, whose binary is registered (loaded) by a node running `avalanchego`, against a user-defined **vmID** (binary file name must be vmID). We can build multiple chains with the registered VM. A VM is registered, only when its built binary is present in the `avalanchego/build/plugins/` directory.

Reviewer 3: 1. nit: Seems confusing to word it as "registered (loaded)", prefer just saying "regsitered" or "loaded".
2. VM IDs are generated, not user-provided.
3. Same plugin directory comment as above.

Author: 2. Cool. Will add more explanations to it. VM IDs are generated from user-provided strings, right?

Reviewer: > 2. Cool. Will add more explanations to it. VM IDs are generated from user-provided strings, right?

yes, see https://docs.avax.network/subnets/subnet-cli#subnet-cli-create-vmid

The VM exposes 3 things for interaction -
* **Static and Non-Static Handlers (API)** for making requests to VM methods
* **Service** for implementing handlers or APIs
* **Client** for making calls to the service's APIs

### What are Handlers

Handlers serve the response for the incoming HTTP requests. Handlers can also be wrapped with **gRPC** for efficiently making calls from other services such as avalanchego. Handler helps in creating APIs. VM implements 2 kinds of handlers -

* **Non-Static Handlers** - They help in interacting with blockchains instantiated by the VM. The API's endpoint will be different for different chains. `/ext/bc/[chainID]`
* **Static Handlers** - They help in directly accessing VM. These are optional for reasons such as parsing genesis bytes required to instantiate new blockchains. `/ext/vm/[vmID]`

> ### Registering a VM

Reviewer 3: I thought VM registry was something the user didn't care about (handled by avalanchego under-the-hood). Should this be in this tutorial? I don't see this in the `blobvm` code.

Reviewer: One suggestion: put what users must implement on the top, and use appendix for detailed explanation at the end of the doc.

Reviewer 2: I think the chain manager itself may be better suited to a separate document or at the very least to an appendix/bottom of the doc as @yulin-dong  suggested.

Reviewer 3: This feels like an unnecessary part of the doc imo, I would be fine removing it. We can just tell users that they need to implement the factory interface, but that's already called out in https://docs.avax.network/subnets/create-a-virtual-machine-vm#factory. I feel like letting users know what's going on under-the-hood is unnecessary for people trying to build VMs and will just be more stuff we need to update as the codebase evolves.

While registering a VM, we store its **factory** against the user-defined **vmID**. As the name suggests, a VM factory can create new VM instances. Using the VM's instance, we can initialize it and can create a new chain. Note that, instantiating a VM is different than initializing it. We initialize the instance of a VM with parameters like genesis, database manager, etc. to create a functional chain.

```go
// Registering a factory inside avalanchego
func (m *manager) RegisterFactory(vmID ids.ID, factory Factory) error {
    ...
    m.factories[vmID] = factory
    ...
}
```

### Instantiating a VM

Each VM has a **factory** that is capable of creating new VM instances. A VM can be initialized as a chain along with handlers for accessing it, only when we have the VM's instance. The factory's `New` method shown below, returns the VM's instance to its caller in `avalanchego`.

> ```go
> // Returning a new VM instance from VM's factory
> func (f *Factory) New(*snow.Context) (interface{}, error) { return &vm.VM{}, nil }
> ```
> 
> VM's methods and other parts of the code are made available to `avalanchego` by **registering** its factory and instantiating a new VM using this factory whenever a chain is required to be built. It's generally in the [`factory.go`](https://github.com/ava-labs/blobvm/blob/master/factory.go) file of the VM.

Reviewer 3: ```suggestion
```go
// Returning a new VM instance from VM's factory
func (f *Factory) New(*snow.Context) (interface{}, error) { return &vm.VM{}, nil }
```
[Reference](https://github.com/ava-labs/blobvm/blob/master/factory.go)

A VM's functionality is exposed to `avalanchego` by registering its factory and instantiating a new VM using this factory whenever a chain is required to be built.
```

```go
// Instantiating a new VM from its factory in avalanchego
func (m *manager) buildChain(chainParams ChainParameters, sb Subnet) (*chain, error) {
    ...
    vmFactory, err := m.VMManager.GetFactory(vmID)
    vm, err := vmFactory.New(ctx.Context)
    ...

> }

Reviewer 3: Let's link references to all code so it's easy to see where the example came from.

```

### Initializing a VM

> Chains are functional, only when the instantiated VMs are initialized. Initializing a VM involves setting up the database, block builder, mempool, genesis state, handlers, etc. This will expose the VM's API handlers, start accepting transactions, gossiping them across the network, building blocks, etc.

Reviewer 3: 1. I would prefer to not use "etc" since it implies that the user has some pre-existing knowledge of how VM's work. If they're reading the tutorial, we shouldn't make that assumption.
2. I think it would be more helpful to just explain what each parameter in `VM#Initialize` does/is used for.

Author: Added more details at the bottom.

```go
if err := vm.Initialize(
    ctx.Context,
    vmDBManager,
    genesisData,
    chainConfig.Upgrade,
    chainConfig.Config,
    msgChan,
    fxs,
    sender,
);
```

> You can refer to the implementation of `vm.initialize` in the [BlobVM](https://github.com/ava-labs/blobvm/blob/1fbb655246a2a41861f40d603ad40299b177252a/vm/vm.go#L92) repository.

Reviewer 3: nit: Unsure if we should use a specific commit version as opposed to master@latest. Both have their pros/cons.
- latest's code is always up-to-date, but the docs will get stale as the vm is updated
- a commit's code is stale, but the docs will always be up-to-date

Author: Using master branch.

### Consensus Engine

## Components of a Virtual Machine

We have to implement the following components in our virtual machine. 

* **Transaction** - Transaction structure, initialization, execution etc.
* **Mempool** - Heap (min and max) for keeping pending transactions locally on the node.
* **Network** - Implements gossip of transactions from mempool to the network.
* **Block** - It handles block creation, initialization, verification, parsing, etc.
* **Builder** - Returns the built preferred block by including transactions from the mempool.
* **Block Builder** - Initiates the gossip of new transactions, notifies the engine that new blocks are ready to be built, and also handles next block generation notification.
* **Storage** - To store and retrieve data of the chain's state.
* **Virtual Machine** - Entry point for all the components to orchestrate them according to the consensus engine and API requests.
* **Service** - API handlers for interacting with VM and initialized chain.
* **Client** - It is used for interacting with service APIs.

> * **Factory** - For creating new instances of the virtual machine.

Reviewer 3: IMO we should link specific interfaces that matter here, users might not know how to implement a Transaction/et al.

> ## Transaction Lifecycle

Reviewer 3: I'm a bit skeptical on if we need to go so deep in this tutorial on how avalanche handles your transactions under-the-hood. I feel like user's should just be concerned with how to implement a VM + transaction types and how avalanche handles it should just be a black-box that they don't care about.

Author: Isn't transaction lifecycle defined by a VM? I have used VM components to explain how the transaction is processed within.

Reviewer 2: Transactions are internal to the VM, the consensus engine only cares about blocks, so it's not necessary to go through it. The phrasing in this document suggests transactions are a hard requirement, so we should make sure that the flexibility of the VM interface is clear (or at least that we don't say that it's more rigid than it is).

Author: Got it. Made it clear in the docs.

> Virtual Machine exposes APIs or handlers for users to make direct RPC to service or use the client to interact with the service. Every change on a chain can happen only through **transactions**. Let's see, how a transaction goes through the network to update the chain's state.

Reviewer 3: Not sure why we bold "transactions". We don't always bold variants of the word (ex. transaction later in the paragraph), so I think we can just not bold it entirely.

* User calls `client.IssueRawTx` or directly make RPC to `service`
* `service.IssueRawTx()` is called using handlers

>     * Receive Tx bytes as arguments

Reviewer 3: +1

    apis := map[string]*common.HTTPHandler{}
    public, err := newHandler(Name, &PublicService{vm: vm})
    if err != nil {
        return nil, err
    }
    apis[PublicEndpoint] = public
    return apis, nil
}
```

In the above function, a new RPC server is created with the `PublicService`. It implements all the RPCs supported by the virtual machine. It has access to a VM instance in its structure, that is passed while creating it (see the above function).

```go
type PublicService struct {
	vm *VM
}
```

Let's have a look at the [`IssueRawTx()`](https://github.com/ava-labs/blobvm/blob/1fbb655246a2a41861f40d603ad40299b177252a/vm/public_service.go#L63) service method -

```go
func (svc *PublicService) IssueRawTx(_ *http.Request, args *IssueRawTxArgs, reply *IssueRawTxReply) error {
	tx := new(chain.Transaction)
	if _, err := chain.Unmarshal(args.Tx, tx); err != nil {
		return err
	}

	// otherwise, unexported tx.id field is empty
	if err := tx.Init(svc.vm.genesis); err != nil {
		return err
	}
	reply.TxID = tx.ID()

	errs := svc.vm.Submit(tx)
	if len(errs) == 0 {
		return nil
	}
	if len(errs) == 1 {
		return errs[0]
	}
	return fmt.Errorf("%v", errs)
}

> ```

Reviewer 4: [Fenced code blocks should be surrounded by blank lines](https://github.com/DavidAnson/markdownlint/blob/v0.25.1/doc/Rules.md#md031)

Blobs are small chunks of data. In BlobVM, we divide a file into small blobs and store them as a key-value pair. The key of these small chunks of a file is linked together as the children of the root. This tree is also stored as JSON data against the root's key.

## How to Use a VM

VM modules are served as a binary plugin to the `avalanchego` software. The nodes that want to power their chains using a particular VM, must have the built binary in their `avalanchego`'s `build/plugins` directory (see [here](../nodes/maintain/avalanchego-config-flags.md#--build-dir-string) for more details). There could be multiple VM plugins in this directory.

## High-Level Overview

Virtual Machines are created as a module, whose binary is registered (loaded) by a node running `avalanchego`, against a user-defined **vmID** (binary file name must be vmID). We can build multiple chains with the registered VM. A VM is registered, only when its built binary is present in the `build/plugins/` directory.

The VM exposes 3 things for interaction:

- **Static and Non-Static Handlers (API)** for making requests to VM methods
- **Service** for implementing handlers or APIs

> - **Client** for making calls to the service's APIs

Reviewer 2: 3 things for interaction is strange wording to me, it sounds like this is the only purpose of the VM, but it doesn't include the actual logic for the consensus engine.

Can you re-phrase this to make it specific to the APIs exposed by the VM?

Also, the Client is not actually part of the VM interface, we have just implemented a client in the VM packages for calling the same APIs. I think it makes sense to include it in this tutorial, but we should clarify that it's not part of the VM itself.

Reviewer 2: I think this piece could be under a section such as `Implementing APIs for BlobVM`

func (m *manager) RegisterFactory(vmID ids.ID, factory Factory) error {
    ...
    m.factories[vmID] = factory
    ...
}
```

### Instantiating a VM

Each VM has a **factory** that is capable of creating new VM instances. A VM can be initialized as a chain along with handlers for accessing it, only when we have the VM's instance. The factory's `New` method shown below, returns the VM's instance to its caller in `avalanchego`.

```go
// Returning a new VM instance from VM's factory
func (f *Factory) New(*snow.Context) (interface{}, error) { return &vm.VM{}, nil }
```

VM's methods and other parts of the code are made available to `avalanchego` by **registering** its factory and instantiating a new VM using this factory whenever a chain is required to be built. It's generally in the [`factory.go`](https://github.com/ava-labs/blobvm/blob/master/factory.go) file of the VM.

```go
// Instantiating a new VM from its factory in avalanchego
func (m *manager) buildChain(chainParams ChainParameters, sb Subnet) (*chain, error) {
    ...
    vmFactory, err := m.VMManager.GetFactory(vmID)
    vm, err := vmFactory.New(ctx.Context)
    ...
}
```

### Initializing a VM

> Chains are functional, only when the instantiated VMs are initialized. Initializing a VM involves setting up the database, block builder, mempool, genesis state, handlers, etc. This will expose the VM's API handlers, start accepting transactions, gossiping them across the network, building blocks, etc.

Reviewer 2: Agree with @joshua-kim . I think a high level explainer of the tasks that a VM needs to fulfill and then subsections for each of those things that get straight to what the parameters are and what it needs to do would be good.

```go
if err := vm.Initialize(
    ctx.Context,
    vmDBManager,
    genesisData,
    chainConfig.Upgrade,
    chainConfig.Config,
    msgChan,
    fxs,
    sender,
);
```

You can refer to the implementation of `vm.initialize` in the [BlobVM](https://github.com/ava-labs/blobvm/blob/1fbb655246a2a41861f40d603ad40299b177252a/vm/vm.go#L92) repository.

### Interaction

Once a VM is initialized, it becomes a functional chain. We can use the handlers to issue transactions, query chain state, and whatever functionalities are being provided by the VM.

> ### Consensus Engine

Reviewer: This is just a header with no contents. Do you intent to add contents under it?

> Once a VM is initialized, it becomes a functional chain. We can use the handlers to issue transactions, query chain state, and whatever functionalities are being provided by the VM.

Reviewer 2: This phrase once a VM is initialized it becomes a functional chain is repeated a few places, but it seems pretty vague to me. Additionally, the VM is initialized before being registered to the chain manager and isn't that functional until bootstrapping is finished, so I think it would be better to be more specific here.

## Components of a Virtual Machine

> We have to implement the following components in our virtual machine.

Reviewer: what does `our virtual machine` refer to?

Author: The VM we want to build.

Reviewer: Then I would suggest to change to the following
```suggestion
We have to implement the following components in a new virtual machine.
```

Author: Nice. Thanks

- **Block** - It handles block creation, initialization, verification, parsing, etc.
- **Builder** - Returns the built preferred block by including transactions from the mempool.
- **Block Builder** - Initiates the gossip of new transactions, notifies the engine that new blocks are ready to be built, and also handles next block generation notification.
- **Storage** - To store and retrieve data of the chain's state.
- **Virtual Machine** - Entry point for all the components to orchestrate them according to the consensus engine and API requests.
- **Service** - API handlers for interacting with VM and initialized chain.
- **Client** - It is used for interacting with service APIs.
- **Factory** - For creating new instances of the virtual machine.

## Transaction Lifecycle

Virtual Machine exposes APIs or handlers for users to make direct RPC to service or use the client to interact with the service. Every change on a chain can happen only through **transactions**. Let's see, how a transaction goes through the network to update the chain's state.

- User calls `client.IssueRawTx` or directly make RPC to `service`
- `service.IssueRawTx()` is called using handlers

>   - Receive Tx bytes as arguments

Reviewer: Should Tx be extended to `transaction` in text?

- Submit transaction object to VM.

Similarly, all other services are implemented.

### Mempool

[Mempool](https://github.com/ava-labs/blobvm/blob/master/mempool/mempool.go) is temporary memory for storing pending transactions. These are maintained at the local node level and can be flushed out when the node restarts. The new transactions received directly from a client or through app gossip, are stored inside this mempool.

Mempool is ideally a Max Heap that pushes new transactions into the heap according to the transaction's price. It keeps the transaction with the highest price at the top. Due to separate implementation of transaction heap and other complexities it has its own `mempool` directory.

Mempool is [initialized](https://github.com/ava-labs/blobvm/blob/master/vm/vm.go#L151) during the initialization of VM.

```go
vm.mempool = mempool.New(vm.genesis, vm.config.MempoolSize)
```

Whenever a transaction is submitted to VM, it first gets initialized, verified, and executed locally. If the transaction looks valid, then it is [added](https://github.com/ava-labs/blobvm/blob/master/vm/vm.go#L431) to mempool.

```go
vm.mempool.Add(tx)

> ```

Reviewer 4: MD031/blanks-around-fences: Fenced code blocks should be surrounded by blank linesmarkdownlint[MD031](https://github.com/DavidAnson/markdownlint/blob/v0.25.1/doc/Rules.md#md031)

```go
if err := b.init(); err != nil {
    return nil, err
}
_, _, err = b.verify()
```

Finally, it will return the block to the caller, which is the VM and hence the consensus engine.

### Storage

[Storage](https://github.com/ava-labs/blobvm/blob/master/chain/storage.go) handles all the database operations like storing transactions, blocks, account balance etc. Everything is stored as key-value pair for all types (block, transaction, balance etc.) of data.

We prefix different types of key with a unique byte. For eg. block identifier is prefixed with `0x0` and transaction ID with `0x1`. Similarly we have [prefix](https://github.com/ava-labs/blobvm/blob/master/chain/storage.go#L56) for other types as well. The prefix and original ID is separated by a `ByteDelimiter`. Prefixing is necessary for identifying the type of raw byte a paticular key is pointing to.

```go
const (
	blockPrefix   = 0x0
	txPrefix      = 0x1
	txValuePrefix = 0x2
	keyPrefix     = 0x3
	balancePrefix = 0x4

	linkedTxLRUSize = 512

	ByteDelimiter byte = '/'
)
```

We have 5 types of key-value pairs in total -

> * **Block** - For storing block data.

Reviewer 4: MD004/ul-style: Unordered list style [Expected: dash; Actual: asterisk]markdownlint[MD004](https://github.com/DavidAnson/markdownlint/blob/v0.25.1/doc/Rules.md#md004)

MD032/blanks-around-lists: Lists should be surrounded by blank linesmarkdownlint[MD032](https://github.com/DavidAnson/markdownlint/blob/v0.25.1/doc/Rules.md#md032)

Let's have a closer look at these functions:

- **SetBalance** - This function will set the balance of an address that will be passed as argument. This is called [internally](https://github.com/ava-labs/blobvm/blob/master/chain/storage.go#L336) by the `ModifyBalance()` function and while [allocating](https://github.com/ava-labs/blobvm/blob/master/chain/genesis.go#L115) the coins to airdrop addresses while creating the genesis block.

```go
func SetBalance(db database.KeyValueWriter, address common.Address, bal uint64) error {
	k := PrefixBalanceKey(address)
	b := make([]byte, 8)
	binary.BigEndian.PutUint64(b, bal)
	return db.Put(k, b)
}
```

- **ModifyBalance** - This function will modify the balance of an address depending upon the parameters `add` and `change`. It will simply perform the `SafeAdd` and `SafeSub` operation on the existing balance. `add` bool parameter indicates whether to add or to subtract the `change` amount from the existing balance. Finally it will call the `SetBalance` function with the updated amount.

>     It is called while executing the transactions. For transfer transactions it will [reduce](https://github.com/ava-labs/blobvm/blob/master/chain/transfer_tx.go#L39) the sender's balance and [increase](https://github.com/ava-labs/blobvm/blob/master/chain/transfer_tx.go#L42) the receiver's balance. And for every transaction it will be called to [reduce](https://github.com/ava-labs/blobvm/blob/master/chain/tx.go#L104) the fee from the sender.

Reviewer 4: Remove the spaces at the beginning of this paragraph.
